### PR TITLE
[filebeat][s3] Add custom parsing script for S3 notifications

### LINKED ISSUE #28828: [Filebeat] aws-s3 - Add validation of S3 Event Notification JSON received from SQS
**Describe the enhancement:**

The aws-s3 input can receive JSON that does not contain `Records` and no error is returned. When this happens the reader assumes the SQS message contained zero object created notifications and carries on normally by deleted the message.

This should be treated as an error because the message is not a valid S3 Event Notification. We should change the struct definition that we use to parse the message such that we can validate that `Records` exists.

https://github.com/elastic/beats/blob/987250d475a88279782c00ec860acdf7f34c3741/x-pack/filebeat/input/awss3/sqs_s3_event.go#L186-L187

**Describe a specific use case for the enhancement or feature:**

The input will be able to better detect error conditions and provide useful error messages.

### PR DESCRIPTION
**Pull request by marc-gr**
## What does this PR do?

Adds the ability to define a custom parsing script to deal with non-standard S3 notifications received through SQS.
Improves error handling for malformed S3 notifications.

## Why is it important?

In some scenarios the received SQS notification might be non-standard. This change lets the user define a custom parsing script to deal with them.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

Closes #28828

### CODE REVIEW

**Comment by leehinman** [comment]
Do we have a use case where the notification received via SQS isn't JSON?

I'm just wondering if we need full JavaScript support or if greater flexibility in JSON parsing would be sufficient.

File: x-pack/filebeat/input/awss3/config.go [hunk 1]
```diff
@@ -151,6 +152,36 @@ func (rc *readerConfig) Validate() error {
 	return nil
 }
 
+type scriptConfig struct {
+	Source            string                 `config:"source"`                               // Inline script to execute.
+	File              string                 `config:"file"`                                 // Source file.
+	Files             []string               `config:"files"`                                // Multiple source files.
```
Comment (efd6) [inline x-pack/filebeat/input/awss3/config.go:158]: Maybe document the order of script source selection; Source>File>Files. Also in input-aws-s3.asciidoc.

Comment (marc-gr) [reply x-pack/filebeat/input/awss3/config.go:158]: There is a comment in the function about only one being allowed at the same time (so no preference, if there are more than one an error is shown). But this was lacking in the docs so I added it there, also.

File: x-pack/filebeat/input/awss3/sqs_s3_event.go [hunk 2]
```diff
@@ -185,6 +188,17 @@ func (p *sqsS3EventProcessor) keepalive(ctx context.Context, log *logp.Logger, w
 }
 
 func (p *sqsS3EventProcessor) getS3Notifications(body string) ([]s3EventV2, error) {
+	// Check if a parsing script is defined. If so, it takes precedence over
+	// format autodetection.
+	if p.script != nil {
+		var raw common.MapStr
+		dec := json.NewDecoder(strings.NewReader(body))
```
Comment (andrewkroh) [inline x-pack/filebeat/input/awss3/sqs_s3_event.go:195]: I think SQS messages can take the form of XML and unstructured text (double check that). Maybe we should pass the raw string through and do the JSON decode in the `parse` implementation if it's required.

Comment (marc-gr) [reply x-pack/filebeat/input/awss3/sqs_s3_event.go:195]: Good point 👍

**Review by andrewkroh — APPROVED** [review]
Nice work! Now this input should be able to handle just about any notification format users throw at it.

**Pull request merged**